File: benchmarks/bench_engine.py
#!/usr/bin/env python3
"""
Python benchmark script for comparing performance with Python's re module.
This mirrors the benchmarks/bench_engine.mojo file for direct comparison.
"""

import re
import time
from typing import Callable


# ===-----------------------------------------------------------------------===#
# Benchmark Infrastructure
# ===-----------------------------------------------------------------------===#


class Benchmark:
    """Simple benchmark infrastructure to mirror Mojo's benchmark system."""

    def __init__(self, num_repetitions: int = 5):
        self.num_repetitions = num_repetitions
        self.results = {}
        self.iterations = {}

    def bench_function(
        self, name: str, fn: Callable[[], None], warmup_iterations: int = 3
    ):
        """Benchmark a function with warmup and multiple repetitions."""
        # Warmup silently
        for _ in range(warmup_iterations):
            fn()

        # Determine target runtime and calculate iterations
        target_runtime = 100_000_000  # 100ms target

        # Run the actual benchmark
        total_time = 0
        actual_iterations = 0

        while total_time < target_runtime and actual_iterations < 100_000:
            start_time = time.perf_counter_ns()
            fn()
            end_time = time.perf_counter_ns()
            fn_time = end_time - start_time
            # print(f"Function {name} took {fn_time} ns")
            total_time += fn_time
            actual_iterations += 1

        # Calculate mean time per run
        mean_time = total_time / actual_iterations

        # Get internal iteration count for accurate per-operation timing
        iterations_map = {
            "literal_match": 100,
            "wildcard_match": 50,
            "range_": 50,
            "anchor_": 100,
            "alternation_": 50,
            "group_": 50,
            "match_all_": 10,
            "complex_email": 2,
            "complex_number": 25,
            "simd_": 10,
            "literal_prefix_": 1,  # Already optimized in Python's re
            "required_literal_": 1,
            "no_literal_baseline": 1,
            "alternation_common_prefix": 1,
        }

        internal_iterations = 1
        for prefix, iters in iterations_map.items():
            if name.startswith(prefix):
                internal_iterations = iters
                break

        # Store results
        self.results[name] = mean_time / internal_iterations
        self.iterations[name] = actual_iterations * internal_iterations

    def dump_report(self):
        """Print summary report of all benchmarks in Mojo format."""
        print("\n=== Benchmark Results ===")
        print("| name                      | met (ms)              | iters  |")
        print("|---------------------------|-----------------------|--------|")

        for name in self.results:
            # Format time in milliseconds with proper precision
            time_ms = self.results[name] / 1_000_000
            iters = self.iterations[name]

            # Right-align values to match Mojo format
            print(f"| {name:<25} | {time_ms:>21.17f} | {iters:>6} |")


# ===-----------------------------------------------------------------------===#
# Benchmark Data Generation
# ===-----------------------------------------------------------------------===#

# Literal optimization test texts
SHORT_TEXT = "hello world this is a test with hello again and hello there"
MEDIUM_TEXT = SHORT_TEXT * 10
LONG_TEXT = SHORT_TEXT * 100
EMAIL_TEXT = "test@example.com user@test.org admin@example.com support@example.com no-reply@example.com"
EMAIL_LONG = EMAIL_TEXT * 5


def make_test_string(length: int, pattern: str = "abcdefghijklmnopqrstuvwxyz") -> str:
    """Generate a test string of specified length by repeating a pattern."""
    if length <= 0:
        return ""

    pattern_len = len(pattern)
    full_repeats = length // pattern_len
    remainder = length % pattern_len

    result = pattern * full_repeats + pattern[:remainder]
    return result


# ===-----------------------------------------------------------------------===#
# Basic Literal Matching Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_literal_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark literal string matching."""

    def benchmark_fn():
        for _ in range(100):
            result = re.search(pattern, test_text)
            # Keep result to prevent optimization
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Wildcard and Quantifier Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_wildcard_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark wildcard and quantifier patterns."""

    def benchmark_fn():
        for _ in range(50):
            result = re.match(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Character Range Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_range_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark character range patterns."""

    def benchmark_fn():
        for _ in range(50):
            result = re.match(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Anchor Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_anchor_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark anchor patterns (^ and $)."""

    def benchmark_fn():
        for _ in range(100):
            result = re.match(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Alternation (OR) Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_alternation_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark alternation patterns (|)."""

    def benchmark_fn():
        for _ in range(50):
            result = re.search(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Group Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_group_match(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark group patterns with quantifiers."""

    def benchmark_fn():
        for _ in range(50):
            result = re.search(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Match All Benchmarks (Global Matching)
# ===-----------------------------------------------------------------------===#


def bench_match_all(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark finding all matches in text."""

    def benchmark_fn():
        for _ in range(10):  # Fewer iterations since findall is more expensive
            results = re.findall(pattern, test_text)
            len(results)  # Keep result

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Complex Pattern Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_complex_email_match(email_text: str) -> Callable[[], None]:
    """Benchmark complex email validation pattern."""
    pattern = r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"

    def benchmark_fn():
        for _ in range(2):
            results = re.findall(pattern, email_text)
            len(results)

    return benchmark_fn


def bench_complex_number_extraction(number_text: str) -> Callable[[], None]:
    """Benchmark extracting numbers from text."""
    pattern = r"\d+\.?\d*"

    def benchmark_fn():
        for _ in range(25):
            results = re.findall(pattern, number_text)
            len(results)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# SIMD-Heavy Character Filtering Benchmarks
# ===-----------------------------------------------------------------------===#


def make_mixed_content_text(length: int) -> str:
    """Generate large mixed content text optimal for SIMD character class testing.

    Args:
        length: The desired length of the test string.

    Returns:
        String with mixed alphanumeric, punctuation, and whitespace content.
    """
    if length <= 0:
        return ""

    # Pattern that creates realistic mixed content with plenty of alphanumeric sequences
    base_pattern = "User123 sent email to user456@domain.com with ID abc789! Status: ACTIVE_2024 (priority=HIGH). "
    pattern_len = len(base_pattern)
    full_repeats = length // pattern_len
    remainder = length % pattern_len

    result = base_pattern * full_repeats + base_pattern[:remainder]
    return result


def bench_simd_heavy_filtering(test_text: str, pattern: str) -> Callable[[], None]:
    """Benchmark SIMD-optimized character class filtering on large mixed content.

    This benchmark is designed to show maximum SIMD performance benefits by:
    - Using character classes that benefit from SIMD lookup tables
    - Processing large amounts of text to amortize SIMD setup costs
    - Using realistic mixed content with alphanumeric sequences
    """

    def benchmark_fn():
        for _ in range(10):  # Fewer iterations due to large text size
            result = re.match(pattern, test_text)
            bool(result)

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Literal Optimization Benchmarks
# ===-----------------------------------------------------------------------===#


def bench_literal_optimization(
    test_text: str, pattern: str, iterations: int
) -> Callable[[], None]:
    """Benchmark literal optimization patterns."""

    def benchmark_fn():
        for _ in range(iterations):
            results = re.findall(pattern, test_text)
            len(results)  # Keep result

    return benchmark_fn


# ===-----------------------------------------------------------------------===#
# Benchmark Main
# ===-----------------------------------------------------------------------===#


def main():
    """Run all benchmarks and display results."""
    m = Benchmark(num_repetitions=5)

    # Pre-create test strings to avoid measurement overhead
    text_1000 = make_test_string(1000)
    text_10000 = make_test_string(10000)
    text_range_1000 = make_test_string(1000, "abc123XYZ")
    text_alternation_1000 = make_test_string(1000, "abcdefghijklmnopqrstuvwxyz")
    text_group_1000 = make_test_string(1000, "abcabcabc")

    # Create complex text patterns
    base_text = make_test_string(100)
    emails = " user@example.com more text john@test.org "
    email_text = f"{base_text} {emails} {base_text} {emails} {base_text}"

    base_number_text = make_test_string(500, "abc def ghi ")
    number_text = (
        f"{base_number_text} 123 price $456.78 quantity 789 {base_number_text}"
    )

    # Basic literal matching
    m.bench_function(
        "literal_match_short",
        bench_literal_match(
            text_1000 + " hello world" + text_1000,  # Ensure "hello" is present
            "hello",
        ),
    )
    m.bench_function(
        "literal_match_long",
        bench_literal_match(
            text_10000 + " hello world" + text_1000,  # Ensure "hello" is present
            "hello",
        ),
    )

    # Wildcard and quantifiers
    m.bench_function("wildcard_match_any", bench_wildcard_match(text_1000, ".*"))
    m.bench_function("quantifier_zero_or_more", bench_wildcard_match(text_1000, "a*"))
    m.bench_function("quantifier_one_or_more", bench_wildcard_match(text_1000, "a+"))
    m.bench_function("quantifier_zero_or_one", bench_wildcard_match(text_1000, "a?"))

    # Character ranges
    m.bench_function("range_lowercase", bench_range_match(text_range_1000, "[a-z]+"))
    m.bench_function("range_digits", bench_range_match(text_range_1000, "[0-9]+"))
    m.bench_function(
        "range_alphanumeric", bench_range_match(text_range_1000, "[a-zA-Z0-9]+")
    )

    # Anchors
    m.bench_function("anchor_start", bench_anchor_match(text_1000, "^abc"))
    m.bench_function("anchor_end", bench_anchor_match(text_1000, "xyz$"))

    # Alternation
    m.bench_function(
        "alternation_simple", bench_alternation_match(text_alternation_1000, "a|b|c")
    )
    m.bench_function(
        "alternation_words",
        bench_alternation_match(text_alternation_1000, "abc|def|ghi"),
    )

    # Groups
    m.bench_function("group_quantified", bench_group_match(text_group_1000, "(abc)+"))
    m.bench_function("group_alternation", bench_group_match(text_group_1000, "(a|b)*"))

    # Global matching (unique to our implementation)
    m.bench_function("match_all_simple", bench_match_all(text_1000, "a"))
    m.bench_function("match_all_pattern", bench_match_all(text_1000, "[a-z]+"))

    # Complex real-world patterns
    m.bench_function("complex_email_extraction", bench_complex_email_match(email_text))
    m.bench_function(
        "complex_number_extraction", bench_complex_number_extraction(number_text)
    )

    # SIMD-Heavy Character Filtering (designed to show maximum SIMD benefit in Mojo comparison)
    large_mixed_text = make_mixed_content_text(10000)
    xlarge_mixed_text = make_mixed_content_text(50000)

    m.bench_function(
        "simd_alphanumeric_large",
        bench_simd_heavy_filtering(large_mixed_text, r"[a-zA-Z0-9]+"),
    )
    m.bench_function(
        "simd_alphanumeric_xlarge",
        bench_simd_heavy_filtering(xlarge_mixed_text, r"[a-zA-Z0-9]+"),
    )
    m.bench_function(
        "simd_negated_alphanumeric",
        bench_simd_heavy_filtering(large_mixed_text, r"[^a-zA-Z0-9]+"),
    )
    m.bench_function(
        "simd_multi_char_class",
        bench_simd_heavy_filtering(large_mixed_text, r"[a-z]+[0-9]+"),
    )

    # Literal Optimization Benchmarks
    m.bench_function(
        "literal_prefix_short",
        bench_literal_optimization(SHORT_TEXT, r"hello.*world", 1),
    )
    m.bench_function(
        "literal_prefix_medium", bench_literal_optimization(MEDIUM_TEXT, r"hello.*", 1)
    )
    m.bench_function(
        "literal_prefix_long", bench_literal_optimization(LONG_TEXT, r"hello.*", 1)
    )
    m.bench_function(
        "required_literal_short",
        bench_literal_optimization(EMAIL_TEXT, r".*@example\.com", 1),
    )
    m.bench_function(
        "required_literal_long",
        bench_literal_optimization(EMAIL_LONG, r".*@example\.com", 1),
    )
    m.bench_function(
        "no_literal_baseline", bench_literal_optimization(MEDIUM_TEXT, r"[a-z]+", 1)
    )
    m.bench_function(
        "alternation_common_prefix",
        bench_literal_optimization(MEDIUM_TEXT, r"(hello|help|helicopter)", 1),
    )

    # Results summary
    m.dump_report()


if __name__ == "__main__":
    main()

File: benchmarks/rust/src/bench_engine.rs
#!/usr/bin/env -S cargo +stable run --release --bin bench_engine --
//! Rust regex benchmark program
//! Mirrors benchmarks/bench_engine.py for direct performance comparison

use mojo_regex_rust_bench::*;
use regex::Regex;
use std::collections::HashMap;
use std::hint::black_box;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let timer = BenchmarkTimer::new();
    let mut all_results = HashMap::new();

    println!("=== RUST REGEX BENCHMARKS ===");
    println!();

    // Pre-create test strings to avoid measurement overhead - scaled up to match Mojo benchmarks
    let text_10000 = make_test_string(10000, "abcdefghijklmnopqrstuvwxyz");     // Increased from 1000 to 10000
    let text_100000 = make_test_string(100000, "abcdefghijklmnopqrstuvwxyz");   // Increased from 10000 to 100000
    let text_range_10000 = make_test_string(10000, "abc123XYZ");
    let text_alternation_10000 = make_test_string(10000, "abcdefghijklmnopqrstuvwxyz");
    let text_group_10000 = make_test_string(10000, "abcabcabc");

    // Create complex text patterns - scaled up to match Mojo benchmarks
    let base_text = make_test_string(2000, "abcdefghijklmnopqrstuvwxyz");  // Increased from 100 to 2000
    let emails = " user@example.com more text john@test.org ";
    let email_text = format!("{} {} {} {} {}", base_text, emails, base_text, emails, base_text);

    let base_number_text = make_test_string(20000, "abc def ghi ");  // Increased from 500 to 20000
    let number_text = format!("{} 123 price $456.78 quantity 789 {}", base_number_text, base_number_text);

    // Pre-compile all regex patterns to avoid compilation overhead in benchmarks
    let patterns = create_all_patterns()?;

    // ===-----------------------------------------------------------------------===
    // Basic Literal Matching Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Literal Matching Benchmarks ===");

    run_benchmark(
        &timer,
        &mut all_results,
        "literal_match_short",
        &patterns.hello,
        &format!("{} hello world {}", text_10000, text_10000),  // Updated variable names
        2000,  // Increased from 100 to 2000
        BenchType::Search
    );

    run_benchmark(
        &timer,
        &mut all_results,
        "literal_match_long",
        &patterns.hello,
        &format!("{} hello world {}", text_100000, text_10000),  // Updated variable names
        2000,  // Increased from 100 to 2000
        BenchType::Search
    );

    // ===-----------------------------------------------------------------------===
    // Wildcard and Quantifier Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Wildcard and Quantifier Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "wildcard_match_any", &patterns.dot_star, &text_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "quantifier_zero_or_more", &patterns.a_star, &text_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "quantifier_one_or_more", &patterns.a_plus, &text_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "quantifier_zero_or_one", &patterns.a_question, &text_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)

    // ===-----------------------------------------------------------------------===
    // Character Range Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Character Range Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "range_lowercase", &patterns.range_a_z, &text_range_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "range_digits", &patterns.range_0_9, &text_range_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "range_alphanumeric", &patterns.range_alnum, &text_range_10000, 1000, BenchType::IsMatch);  // Updated text size and iterations (50->1000)

    // ===-----------------------------------------------------------------------===
    // Anchor Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Anchor Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "anchor_start", &patterns.anchor_start, &text_10000, 2000, BenchType::IsMatch);  // Updated text size and iterations (100->2000)
    run_benchmark(&timer, &mut all_results, "anchor_end", &patterns.anchor_end, &text_10000, 2000, BenchType::IsMatch);  // Updated text size and iterations (100->2000)

    // ===-----------------------------------------------------------------------===
    // Alternation Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Alternation Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "alternation_simple", &patterns.alt_simple, &text_alternation_10000, 1000, BenchType::Search);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "alternation_words", &patterns.alt_words, &text_alternation_10000, 1000, BenchType::Search);  // Updated text size and iterations (50->1000)

    // ===-----------------------------------------------------------------------===
    // Group Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Group Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "group_quantified", &patterns.group_quantified, &text_group_10000, 1000, BenchType::Search);  // Updated text size and iterations (50->1000)
    run_benchmark(&timer, &mut all_results, "group_alternation", &patterns.group_alternation, &text_group_10000, 1000, BenchType::Search);  // Updated text size and iterations (50->1000)

    // ===-----------------------------------------------------------------------===
    // NEW: Optimization Showcase Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Optimization Showcase Benchmarks ===");

    // Test case 1: Large alternation (8 branches) - benefits from increased branch limit (3→8)
    let fruit_text = "I love eating apple and banana and cherry and date and elderberry and fig and grape with honey";
    run_benchmark(&timer, &mut all_results, "large_alternation_8_branches", &patterns.large_alternation, fruit_text, 1000, BenchType::Search);

    // Test case 2: Deeply nested groups (depth 4) - benefits from increased depth tolerance (3→4)
    let nested_text = "Testing deep nested patterns with abcdefgh characters";
    run_benchmark(&timer, &mut all_results, "deep_nested_groups_depth4", &patterns.deep_nested, nested_text, 1000, BenchType::Search);

    // Test case 3: Literal-heavy alternation - benefits from 80% threshold detection
    let user_text = "Login attempts: user123 failed, admin456 success, guest789 failed, root000 success";
    run_benchmark(&timer, &mut all_results, "literal_heavy_alternation", &patterns.literal_heavy, user_text, 1000, BenchType::Search);

    // Test case 4: Complex group with 5 children - benefits from increased children limit (3→5)
    let mixed_text = "Found: hello123ab, world456cd, test789ef, demo012gh, sample345ij in the data";
    run_benchmark(&timer, &mut all_results, "complex_group_5_children", &patterns.complex_group, mixed_text, 1000, BenchType::Search);

    // ===-----------------------------------------------------------------------===
    // Global Matching Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Global Matching Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "match_all_simple", &patterns.a, &text_10000, 200, BenchType::FindAll);  // Updated text size and iterations (10->200)
    run_benchmark(&timer, &mut all_results, "match_all_pattern", &patterns.range_a_z, &text_10000, 200, BenchType::FindAll);  // Updated text size and iterations (10->200)

    // ===-----------------------------------------------------------------------===
    // Complex Pattern Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Complex Pattern Benchmarks ===");

    run_benchmark(&timer, &mut all_results, "complex_email_extraction", &patterns.email, &email_text, 40, BenchType::FindAll);  // Increased from 2 to 40
    run_benchmark(&timer, &mut all_results, "complex_number_extraction", &patterns.number, &number_text, 500, BenchType::FindAll);  // Increased from 25 to 500

    // ===-----------------------------------------------------------------------===
    // SIMD-Optimized Character Filtering Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== SIMD-Optimized Character Filtering Benchmarks ===");

    let large_mixed_text = make_mixed_content_text(100000);  // Increased from 10000 to 100000
    let xlarge_mixed_text = make_mixed_content_text(500000);  // Increased from 50000 to 500000

    run_benchmark(&timer, &mut all_results, "simd_alphanumeric_large", &patterns.range_alnum, &large_mixed_text, 200, BenchType::IsMatch);  // Increased from 10 to 200
    run_benchmark(&timer, &mut all_results, "simd_alphanumeric_xlarge", &patterns.range_alnum, &xlarge_mixed_text, 200, BenchType::IsMatch);  // Increased from 10 to 200
    run_benchmark(&timer, &mut all_results, "simd_negated_alphanumeric", &patterns.negated_alnum, &large_mixed_text, 200, BenchType::IsMatch);  // Increased from 10 to 200
    run_benchmark(&timer, &mut all_results, "simd_multi_char_class", &patterns.multi_char_class, &large_mixed_text, 200, BenchType::IsMatch);  // Increased from 10 to 200

    // ===-----------------------------------------------------------------------===
    // Literal Optimization Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== Literal Optimization Benchmarks ===");

    let medium_text = get_medium_text();
    let long_text = get_long_text();
    let email_long = get_email_long();

    run_benchmark(&timer, &mut all_results, "literal_prefix_short", &patterns.literal_prefix_short, SHORT_TEXT, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "literal_prefix_medium", &patterns.literal_prefix_medium, &medium_text, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "literal_prefix_long", &patterns.literal_prefix_long, &long_text, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "required_literal_short", &patterns.required_literal, EMAIL_TEXT, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "required_literal_long", &patterns.required_literal, &email_long, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "no_literal_baseline", &patterns.range_a_z, &medium_text, 1, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "alternation_common_prefix", &patterns.alt_common_prefix, &medium_text, 1, BenchType::FindAll);

    // ===-----------------------------------------------------------------------===
    // US Phone Number Benchmarks
    // ===-----------------------------------------------------------------------===
    println!("=== US Phone Number Benchmarks ===");

    let phone_text = make_phone_test_data(1000);

    run_benchmark(&timer, &mut all_results, "simple_phone", &patterns.simple_phone, &phone_text, 100, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "flexible_phone", &patterns.flexible_phone, &phone_text, 100, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "multi_format_phone", &patterns.multi_format_phone, &phone_text, 50, BenchType::FindAll);
    run_benchmark(&timer, &mut all_results, "phone_validation", &patterns.phone_validation, "555-123-4567", 500, BenchType::IsMatch);

    // ===-----------------------------------------------------------------------===
    // Results Summary
    // ===-----------------------------------------------------------------------===
    println!("\n=== Benchmark Results ===");
    print_results_table(&all_results);

    // Export to JSON
    export_json_results(&all_results)?;

    Ok(())
}

#[derive(Debug)]
enum BenchType {
    IsMatch,
    Search,
    FindAll,
}

struct CompiledPatterns {
    hello: Regex,
    dot_star: Regex,
    a_star: Regex,
    a_plus: Regex,
    a_question: Regex,
    a: Regex,
    range_a_z: Regex,
    range_0_9: Regex,
    range_alnum: Regex,
    negated_alnum: Regex,
    multi_char_class: Regex,
    anchor_start: Regex,
    anchor_end: Regex,
    alt_simple: Regex,
    alt_words: Regex,
    alt_common_prefix: Regex,
    group_quantified: Regex,
    group_alternation: Regex,
    email: Regex,
    number: Regex,
    literal_prefix_short: Regex,
    literal_prefix_medium: Regex,
    literal_prefix_long: Regex,
    required_literal: Regex,
    simple_phone: Regex,
    flexible_phone: Regex,
    multi_format_phone: Regex,
    phone_validation: Regex,
    large_alternation: Regex,
    deep_nested: Regex,
    literal_heavy: Regex,
    complex_group: Regex,
}

fn make_phone_test_data(num_phones: usize) -> String {
    let phone_patterns = [
        "555-123-4567",
        "(555) 123-4567",
        "555.123.4567",
        "5551234567",
        "+1-555-123-4567",
        "1-555-123-4568",
        "(555)123-4569",
        "555 123 4570"
    ];
    let filler_text = " Contact us at ";
    let extra_text = " or email support@company.com for assistance. ";

    let mut result = String::new();
    for i in 0..num_phones {
        result.push_str(filler_text);
        let pattern_idx = i % phone_patterns.len();
        result.push_str(phone_patterns[pattern_idx]);
        result.push_str(extra_text);
    }

    result
}

fn create_all_patterns() -> Result<CompiledPatterns, Box<dyn std::error::Error>> {
    Ok(CompiledPatterns {
        hello: Regex::new("hello")?,
        dot_star: Regex::new(".*")?,
        a_star: Regex::new("a*")?,
        a_plus: Regex::new("a+")?,
        a_question: Regex::new("a?")?,
        a: Regex::new("a")?,
        range_a_z: Regex::new("[a-z]+")?,
        range_0_9: Regex::new("[0-9]+")?,
        range_alnum: Regex::new("[a-zA-Z0-9]+")?,
        negated_alnum: Regex::new("[^a-zA-Z0-9]+")?,
        multi_char_class: Regex::new("[a-z]+[0-9]+")?,
        anchor_start: Regex::new("^abc")?,
        anchor_end: Regex::new("xyz$")?,
        alt_simple: Regex::new("a|b|c")?,
        alt_words: Regex::new("abc|def|ghi")?,
        alt_common_prefix: Regex::new("(hello|help|helicopter)")?,
        group_quantified: Regex::new("(abc)+")?,
        group_alternation: Regex::new("(a|b)*")?,
        email: Regex::new(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")?,
        number: Regex::new(r"\d+\.?\d*")?,
        literal_prefix_short: Regex::new("hello.*world")?,
        literal_prefix_medium: Regex::new("hello.*")?,
        literal_prefix_long: Regex::new("hello.*")?,
        required_literal: Regex::new(r".*@example\.com")?,
        simple_phone: Regex::new(r"\d{3}-\d{3}-\d{4}")?,
        flexible_phone: Regex::new(r"\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}")?,
        multi_format_phone: Regex::new(r"\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4}|\d{3}-\d{3}-\d{4}|\d{10}")?,
        phone_validation: Regex::new(r"^\+?1?[\s.-]?\(?([2-9]\d{2})\)?[\s.-]?([2-9]\d{2})[\s.-]?(\d{4})$")?,
        large_alternation: Regex::new(r"(apple|banana|cherry|date|elderberry|fig|grape|honey)")?,
        deep_nested: Regex::new(r"(?:(?:(?:a|b)|(?:c|d))|(?:(?:e|f)|(?:g|h)))")?,
        literal_heavy: Regex::new(r"(user123|admin456|guest789|root000|test111|demo222|sample333|client444)")?,
        complex_group: Regex::new(r"(hello|world|test|demo|sample)[0-9]{3}[a-z]{2}")?,
    })
}

fn run_benchmark(
    timer: &BenchmarkTimer,
    results: &mut HashMap<String, BenchmarkResult>,
    name: &str,
    pattern: &Regex,
    text: &str,
    inner_iterations: usize,
    bench_type: BenchType,
) {
    let result = match bench_type {
        BenchType::IsMatch => {
            timer.bench_function(|| {
                for _ in 0..inner_iterations {
                    black_box(pattern.is_match(black_box(text)));
                }
            })
        }
        BenchType::Search => {
            timer.bench_function(|| {
                for _ in 0..inner_iterations {
                    black_box(pattern.find(black_box(text)));
                }
            })
        }
        BenchType::FindAll => {
            timer.bench_function(|| {
                for _ in 0..inner_iterations {
                    let matches: Vec<_> = pattern.find_iter(black_box(text)).collect();
                    black_box(matches.len());
                }
            })
        }
    };

    // Adjust time per operation by dividing by inner iterations
    let adjusted_result = BenchmarkResult {
        time_ns: result.time_ns / inner_iterations as f64,
        time_ms: result.time_ms / inner_iterations as f64,
        iterations: result.iterations * inner_iterations as u64,
    };

    results.insert(name.to_string(), adjusted_result);
    println!("✓ {}", name);
}

fn print_results_table(results: &HashMap<String, BenchmarkResult>) {
    println!("| name                      | met (ms)              | iters  |");
    println!("|---------------------------|-----------------------|--------|");

    let mut sorted_results: Vec<_> = results.iter().collect();
    sorted_results.sort_by_key(|(name, _)| name.as_str());

    for (name, result) in sorted_results {
        println!("| {:<25} | {:>21.17} | {:>6} |",
                 name, result.time_ms, result.iterations);
    }
}

fn export_json_results(results: &HashMap<String, BenchmarkResult>) -> Result<(), Box<dyn std::error::Error>> {
    let benchmark_results = BenchmarkResults {
        engine: "rust".to_string(),
        timestamp: chrono::Utc::now().to_rfc3339(),
        results: results.clone(),
    };

    std::fs::create_dir_all("../results")?;

    let json_content = serde_json::to_string_pretty(&benchmark_results)?;
    std::fs::write("../results/rust_results.json", json_content)?;

    println!("\n=== BENCHMARK COMPLETE ===");
    println!("Results exported to: ../results/rust_results.json");

    Ok(())
}
